lib/spack/spack/test/llnl/util/symlink.py

```diff
@@ -0,0 +1,265 @@
+
+    test_parent_dir = tmpdir_factory.mktemp("symlink_test")
+    assert os.path.exists(test_parent_dir)
+
+    # Run tests
+    yield test_parent_dir
+
+    # Clean up test directory
+    shutil.rmtree(test_parent_dir)
+
+
+def test_symlink_file(stage):
+    """Test the symlink.symlink functionality on all operating systems for a file"""
+    test_dir = tempfile.mkdtemp(dir=stage)
+    fd, real_file = tempfile.mkstemp(prefix="real", suffix=".txt", dir=test_dir)
+    try:
+        assert os.path.exists(real_file)
+        link_file = tempfile.mktemp(prefix="link", suffix=".txt", dir=test_dir)
+        assert os.path.exists(link_file) is False
+        symlink.symlink(source_path=real_file, link_path=link_file)
+        assert os.path.exists(link_file)
+        assert symlink.islink(link_file)
+    finally:
+        os.close(fd)
+
+
+def test_symlink_dir(stage):
+    """Test the symlink.symlink functionality on all operating systems for a directory"""
+    test_dir = tempfile.mkdtemp(dir=stage)
+    real_dir = os.path.join(test_dir, "real_dir")
+    link_dir = os.path.join(test_dir, "link_dir")
+    os.mkdir(real_dir)
+    assert os.path.exists(real_dir)
+    symlink.symlink(source_path=real_dir, link_path=link_dir)
+    assert os.path.exists(link_dir)
+    assert symlink.islink(link_dir)
+
+
+def test_symlink_source_not_exists(stage):
+    """Test the symlink.symlink method for the case where a source path does not exist"""
+    test_dir = tempfile.mkdtemp(dir=stage)
+    real_dir = os.path.join(test_dir, "real_dir")
+    link_dir = os.path.join(test_dir, "link_dir")
+    assert not os.path.exists(real_dir)
+    assert not os.path.exists(link_dir)
+    with pytest.raises(symlink.SymlinkError):
+        symlink.symlink(source_path=real_dir, link_path=link_dir)
+
+
+def test_symlink_src_relative_to_link(stage):
+    """Test the symlink.symlink functionality where the source value exists relative to the link
+    but not relative to the cwd"""
+    test_dir = tempfile.mkdtemp(dir=stage)
+    subdir_1 = os.path.join(test_dir, "a")
+    subdir_2 = os.path.join(subdir_1, "b")
+    link_dir = os.path.join(subdir_1, "c")
+    os.mkdir(subdir_1)
+    os.mkdir(subdir_2)
+    prev_dir = os.getcwd()
+    fd, real_file = tempfile.mkstemp(prefix="real", suffix=".txt", dir=subdir_2)
+    try:
+        os.chdir(test_dir)
+        assert os.path.exists(real_file)
+        link_file = tempfile.mktemp(prefix="link", suffix=".txt", dir=subdir_1)
+        assert os.path.exists(link_file) is False
+        symlink.symlink(
+            source_path=f"b/{os.path.basename(real_file)}",
+            link_path=f"a/{os.path.basename(link_file)}",
+        )
+        assert os.path.exists(link_file)
+        assert symlink.islink(link_file)
+        # Check dirs
+        assert not os.path.lexists(link_dir)
+        symlink.symlink(source_path="b", link_path="a/c")
+        assert os.path.lexists(link_dir)
+    finally:
+        os.chdir(prev_dir)
+        os.close(fd)
```

Comment:
Using the tmpdir fixture is how we generally handle automatic cleanup.

Also, we typically use `with working_dir()` to handle maintaining the working directory.

Reply:
Updated the tests to use tmpdir instead of stage